### DISCUSSION TITLE
Add option-label and option-value to QOptionGroup

### DISCUSSION THREAD

**Discussion by jaalders** [quasarframework/quasar#16874] [Ideas / Proposals] [answered]
Hi there,

I know that technically you can get around the issue of a multi-select option group with <q-select> with a 'multiple' option added to a <q-select>, but if you are trying to use <q-option-group> your options param has to be tightly coupled to label value like so 

```
options: [
{
  label: 'Option 1',
  value: 'op1'
},
```

If I wanted to use an interface for the options array, I will not be able to use it due to the conflict on the options param from Quasar. 

Take this example:

```
// use generics to allow for on the fly naming aka: value/label
export interface IMenuDietaryOptions {
  [key: string]: number | string;
}

// no error here
const dietaryOptions: IMenuDietaryOptions[] = [
  { value: 1, label: 'Dairy Free' },
  { value: 2, label: 'Gluten Free' },
  { value: 3, label: 'Keto' },
  { value: 4, label: 'Nut Free' },
  { value: 5, label: 'Paleo' },
  { value: 6, label: 'Vegan' },
  { value: 7, label: 'Vegetarian' },
];

  <q-option-group
    v-model="menuItem.dietaryOptions"
    :options="dietaryOptions"
    color="primary"
  />
```

The :options section has the following error

```
  Type 'IMenuDietaryOptions[]' is not assignable to type '{ [props: string]: any; label: string; value: any; disable?: boolean | undefined; }[]'.ts(2322)
  index.d.ts(7168, 3): The expected type comes from property 'options' which is declared here on type 'VNodeProps & AllowedComponentProps & ComponentCustomProps & QOptionGroupProps & Record<string, unknown>'
```

Perhaps I am just missing something, and by all means if that is the case, please guide me, but if we were to allow map-options and emit-value to <q-option-group> we could specify our label and value pair and we could keep our type safety

**Comment by yusufkandemir** [comment]
You need to use the correct type, or don't enforce a type at all(_e.g. let TS infer the type_).
```ts
import { QOptionGroupProps } from 'quasar';

const dietaryOptions: QOptionGroupProps['options'] = [
  { value: 1, label: 'Dairy Free' },
  { value: 2, label: 'Gluten Free' },
  { value: 3, label: 'Keto' },
  { value: 4, label: 'Nut Free' },
  { value: 5, label: 'Paleo' },
  { value: 6, label: 'Vegan' },
  { value: 7, label: 'Vegetarian' },
];
// or just
const dietaryOptions = [
  // ...
];
```

**Reply by jaalders** [reply]
While this may work coding-wise, having the benefit of map-options/emit-value for radio groups would allow users to use `<q-option-group>` UI styling and not be forced to use `<q-select>`'s UI styling to achieve this.

**Reply by yusufkandemir** [reply]
Your question was about type safety, so this is not just something "may work coding-wise", this is the answer. Regarding the other request, it should've been stated clearly and should be a separate discussion, but anyway. QOptionGroup already works that way. Its model will not hold the objects themselves, but only the values, so the model will be like `[1, 3, 4]` with those options. So, I am not sure what you mean.

**Reply by jaalders** [reply]
If my comment came off as rude, that was certainly not my intention. I just was just reviewing the documentation and saw map-options/emit-value as a solution to my issue and figured it would be nice to apply it to `<q-option-group>` as well since there was more flexibility with that `<q-select>` type, but didn't allow for a UI of radio buttons.

In regards to a separate discussion. If truly believe that my request has merit. I can certainly make another thread, but it would be great if you could help me choose the right verbiage to convey my intentions.

**Reply by yusufkandemir** [reply]
Your TS problem was somewhat clear, but your other request is still unclear. With the only thing I could understand, I said QOptionGroup already works like `map-options`+`emit-value` behavior of QSelect. So, you should further clarify what you have now and what you want, it's not even clear what is the type of `menuItem.dietaryOptions` and what you want it to be.

**Reply by jaalders** [reply]
`<q-select>` allows you to map both value and label which is great, because if you have your own interface you can specify your label and value from what is in that custom interface, and doesn't require a import from Quasar. However, with `<q-option-group>` you're limited to the label/value naming constraints. If I wanted to use my own custom interface for `<q-option-group>` I wouldn't be able to do so.

**Reply by yusufkandemir** [reply]
Still hard to understand how it's related to `map-options`+`emit-value`. But, I have a guess. You want to have the equivalent of `option-label` and `option-value` to be added to QOptionGroup. So you could do:
```html
<q-option-group
  ...
  :options="options"
  option-label="name"
  option-value="type"
/>

<script setup>
const options = [
  { name: 'Gluten Free', type: 1 }
];
</script>
```
instead of
```html
<q-option-group
  ...
  :options="options"
/>

<script setup>
const myOptions = [
  { name: 'Gluten Free', type: 1 }
];

const options = myOptions.map(({ name, type }) => ({
  label: name,
  value: type
}));
</script>
```
is this what you what?

**Reply by jaalders** [reply]
This is correct

**Reply by RVXD** [reply]
Yes, please, can this be added? It is confusing that the q-select does have 'option-label' and 'option-value' and q-option-group does not have this. If implemented, switching between either dropdown or option group would be a lot easier.

**Comment by rstoenescu** [comment, marked as answer]
Just pushed: feat+perf(QOptionsGroup): new props -> option-value, option-label & option-disable #16874
Will be available in Quasar v2.17.0

**Reply by jaalders** [reply]
Took a little bit to get there, but glad to see it get added! 👍
